src/oneD/IonFlow.cpp

```diff
@@ -0,0 +1,416 @@
+    }
+}
+
+void IonFlow::setElectricVoltage(const double v1, const double v2)
+{
+    // This method can be used when you want to add external voltage
+    m_inletVoltage = v1;
+    m_outletVoltage = v2;
+}
+
+void IonFlow::eval(size_t jg, doublereal* xg,
+                  doublereal* rg, integer* diagg, doublereal rdt)
+{
+    StFlow::eval(jg, xg, rg, diagg, rdt);
+    // start of local part of global arrays
+    doublereal* x = xg + loc();
+    doublereal* rsd = rg + loc();
+    integer* diag = diagg + loc();
+
+    size_t jmin, jmax;
+    if (jg == npos) { // evaluate all points
+        jmin = 0;
+        jmax = m_points - 1;
+    } else { // evaluate points for Jacobian
+        size_t jpt = (jg == 0) ? 0 : jg - firstPoint();
+        jmin = std::max<size_t>(jpt, 1) - 1;
+        jmax = std::min(jpt+1,m_points-1);
+    }
+
+    if (m_phase == 3) {
+        for (size_t j = jmin; j <= jmax; j++) {
+            if (j == 0) {
+                rsd[index(c_offset_P, j)] = m_inletVoltage - phi(x,j);
+                diag[index(c_offset_P, j)] = 0;
+                // set ions boundary for better convergence
+                for (size_t k : m_kCharge) {
+                    rsd[index(c_offset_Y + k, j)] = Y(x,k,j+1) - Y(x,k,j);
+                }
+            } else if (j == m_points - 1) {
+                rsd[index(c_offset_P, j)] = m_outletVoltage - phi(x,j);
+                diag[index(c_offset_P, j)] = 0;
+            } else {
+                evalPoisson(j,x,rsd,diag,rdt);
+                if (!m_do_velocity[j]) {
+                    // This method is used when you diable energy equation
+                    // but still maintain the velocity profile
+                    rsd[index(c_offset_U, j)] = u(x,j) - u_fixed(j);
+                    diag[index(c_offset_U, j)] = 0;
+                }
+            }
+        }
+    }
+}
+
+void IonFlow::evalPoisson(size_t j, doublereal* x, doublereal* rsd, integer* diag, doublereal rdt)
+{
+    //-----------------------------------------------
+    //    Poisson's equation
+    //
+    //    dE/dz = e/eps_0 * sum(q_k*n_k)
+    //
+    //    E = -dV/dz 
+    //-----------------------------------------------
+    double chargeDensity = 0.0;
+    for (size_t k : m_kCharge) {
+        chargeDensity += m_speciesCharge[k] * ElectronCharge * ND(x,k,j);
+    }
+    rsd[index(c_offset_P, j)] = dEdz(x,j) - chargeDensity / epsilon_0;
+    diag[index(c_offset_P, j)] = 0;
+}
+
+void IonFlow::solvePoissonEqn(size_t j)
+{
+    bool changed = false;
+    if (j == npos) {
+        for (size_t i = 0; i < m_points; i++) {
+            if (!m_do_poisson[i]) {
+                changed = true;
+            }
+            m_do_poisson[i] = true;
+        }
+    } else {
+        if (!m_do_poisson[j]) {
+            changed = true;
+        }
+        m_do_poisson[j] = true;
+    }
+    m_refiner->setActive(0, true);
+    m_refiner->setActive(1, true);
+    m_refiner->setActive(2, true);
+    m_refiner->setActive(4, true);
```

Comment:
These should use the `c_offset_P` and friends variables, right? Same goes for the parallel methods.